acvm/src/lib.rs

```diff
@@ -79,18 +87,35 @@ pub trait PartialWitnessGenerator {
                     Opcode::Block(id, trace) => blocks.solve(*id, trace, initial_witness),
                 };
                 match resolution {
-                    Ok(_) => {
-                        // We do nothing in the happy case
+                    Ok(GateResolution::Solved) => {
+                        stalled = false;
                     }
-                    Err(OpcodeResolutionError::OpcodeNotSolvable(_)) => {
-                        // For opcode not solvable errors, we push those opcodes to the back as
+                    Ok(GateResolution::InProgress) => {
+                        stalled = false;
+                        unresolved_opcodes.push(opcode.clone());
+                    }
+                    Ok(GateResolution::Stalled(not_solvable)) => {
+                        if opcode_not_solvable.is_none() {
+                            // we keep track of the first unsolvable opcode
+                            opcode_not_solvable = Some(not_solvable);
+                        }
+                        // We push those opcodes not solvable to the back as
                         // it could be because the opcodes are out of order, i.e. this assignment
                         // relies on a later opcodes' results
                         unresolved_opcodes.push(opcode.clone());
                     }
+                    Err(OpcodeResolutionError::OpcodeNotSolvable(_)) => {
+                        unreachable!("ICE - Result should have been converted to GateResolution")
+                    }
                     Err(err) => return Err(err),
                 }
             }
+            if stalled && !unresolved_opcodes.is_empty() {
+                return Err(OpcodeResolutionError::OpcodeNotSolvable(
+                    opcode_not_solvable
+                        .expect("infallible: cannot be stalled and none at the same time"),
```

Comment:
```suggestion
                        .expect("infallible: cannot be stalled and None at the same time"),
```

Nit but this makes it clearer that we're talking about an actual `None` value.